### ISSUE TITLE
Allow pl-file-upload to accept files with any name

### ISSUE THREAD

**Issue by neetesh-nks** [PrairieLearn/PrairieLearn#3063]
While using pl-file-upload for manual grading. It would be helpful to allow students to upload their work in different file formats without specific naming requirements.

**Comment by echuber2** [comment]
I think that right now, the upload element uses the specific names to distinguish between the received files and determine if the student has uploaded all the required files. (We recently added a feature to accept filenames without case sensitivity, but it still expects the filename to match otherwise.) So there could be an option added to accept files regardless of name, but the preview and other mechanisms will need a lot of editing.

A temporary workaround is to request a zip file, and let the zip file be what has a specific name. Then students can put whatever they want in the zip file.

**Comment by jonatanschroeder** [comment]
This has been a request by several instructors at UBC recently, where they want to allow students to upload any file with a list of possible extensions (jpg, png, pdf, etc.).

Here's a suggested implementation: allow `pl-file-upload` to have specific file rules as sub-elements like:
```
<pl-file-upload>
  <pl-file file-name="submitted.jpeg" accepted-source-name="*.jpg,*.jpeg" min-needed="0" max-needed="1"></pl-file-rule>
  <pl-file file-name="submitted.png" accepted-source-name="*.png" min-needed="0" max-needed="1"></pl-file-rule>
  <pl-file file-name="submitted.pdf" accepted-source-name="*.pdf" min-needed="0" max-needed="1"></pl-file-rule>
</pl-file-upload>
```

The file would then be renamed automatically inside the element according to specific rules. Some details still need to be decided (e.g., what happens if multiple rules apply?), but this could cover a larger variety of cases.

**Comment by echuber2** [comment]
This could be added in a similar way to the edits I made here: https://github.com/PrairieLearn/PrairieLearn/pull/2955

**Comment by eliotwrobson** [comment]
It looks like that file rules feature is robust enough for the developer to work with, but it seems like it requires some substantial thought for how this would be rendered on the frontend. For example, if you have `max-needed="2"` for one of the file extensions in the list, how would the files be renamed? And if the names are different, then how would the student select which file their uploaded file should be renamed to?

Off the top of my head, it seems like the current UX only naturally allows for one rename rule per file extension, and this might be enough to capture most cases. @jonatanschroeder what do you think?

**Comment by jonatanschroeder** [comment]
The most common case we see where this is used is images and PDF files. We have discussed this, and the likely best option is to have a custom element for image/document upload, which would have the advantage of allowing for the implementation of better preview options, the ability to use a camera for input, reordering images in a specific order, etc.

As for the file upload, indeed these rules may be more complicated than needed. Other options may be more viable, such as using string distance to suggest renames, but it may be better to clarify those outside the image context and which use cases they may serve.

**Comment by eliotwrobson** [comment]
I think that tracks, a custom element for uploading images and documents allows for building in better support for just those use cases. Then with file upload, we can focus new features on supporting use cases outside of what the new element would cover.

**Comment by mfbutner** [comment]
One other use case aside from files and images is in my programming classes. Sometimes I don't care what the names of the source files my students submit are as they have to provide a Makefile or CMakeLists.txt that must compile their program into an executable name X.out. Or with python they submit a python package and a file named exactly main.py that is the starting point for their program. 

I have worked around this by having them submit a zip file with their submission but that makes it harder to view in the file previewer and I must download the submissions to do the manual grading. 

I'm not sure if jonatan's suggested format would work for my case.

**Comment by jonatanschroeder** [comment]
> One other use case aside from files and images is in my programming classes. Sometimes I don't care what the names of the source files my students submit are as they have to provide a Makefile or CMakeLists.txt that must compile their program into an executable name X.out. Or with python they submit a python package and a file named exactly main.py that is the starting point for their program.

Yes, neither of my suggestions would work directly with this use-case. In your case you want to accept the files and keep their names, which is not what my original suggestion was about. Maybe some ideas can be taken from that example, though.

**Comment by eliotwrobson** [comment]
I think this would be resolved by https://github.com/PrairieLearn/PrairieLearn/pull/10317
